s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -198,6 +197,32 @@ import scala.util.{Failure, Success, Try}
       }
   }.mapMaterializedValue(_ => NotUsed)
 
+  def getObject(
+      s3Location: S3Location,
+      range: Option[ByteRange],
+      versionId: Option[String],
+      s3Headers: S3Headers
+  ): Source[ByteString, Future[ObjectMetadata]] = {
+    val headers = s3Headers.headersFor(GetObject)
+    val objectMetadataMat = Promise[ObjectMetadata]()
+
+    Source
+      .fromMaterializer { (mat, attr) =>
+        implicit val materializer: Materializer = mat
+        issueRequest(s3Location, rangeOption = range, versionId = versionId, s3Headers = headers)(mat, attr)
+          .map(response => response.withEntity(response.entity.withoutSizeLimit))
+          .mapAsync(parallelism = 1)(entityForSuccess)
+          .map {
+            case (entity, headers) =>
+              objectMetadataMat.success(computeMetaData(headers, entity))
```

Comment:
My only slight concern is when this gets evaluated to complete the `objectMetadataMat` `Promise` due to the lazy nature of akka-streams.

This can be problematic because ideally you want the materialized value to complete ASAP (i.e. as soon as you get the response headers) so that you can fullfill the usecase of reusing the `Source[ByteString,_]` in other stream computations as soon as the `Future[ObjectMetadata]` gets completed (and you **don't** want this to occur when the body has completed streaming)

Reply:
Do you have a more concrete example where you suspect this would happen? It's not obvious to me there is any problem lurking here.

Reply:
What I mentioned earlier is just a hypothetical problem that popped in my head so I mentioned it but in reality it can be a complete non concern in which case you can just ignore it.

Reply:
Ok, in that case I'd expect that it is fine because when materializing the stream there will always be a downstream triggering demand, leading to the request coming in here and completing the promise, if a stream for some reason has backpressure but still want that promise to complete early they can `detach` to trigger the request early.